keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,602 @@
+| :--- | :--- | :--- | :--- | :--- |
+| Single Approach Test | Unit Test | Small/Large number of nodes | Distribute EndpointSlices as expected | |
+| |  | Balanced/Imbalanced endpoints | Distribute EndpointSlices as expected | |
+| |  | Zero endpoints in some zones | Distribute EndpointSlices as expected | |
+| |  | Invalid annotation | Approach sets to default, distribute EndpointSlices as expected | |
+| `Prefer` Test | Unit Test | Invalid starting threshold | Set starting threshold to default and start `Prefer` approach with condition meets| This is applicable for customized starting threshold only|
+| |  | Endpoints below `starting threshold + padding` | `Prefer` approach will not be conducted | |
+| |  | Endpoints above `starting threshold + padding` | `Prefer` works as expected | |
+| Approach Transition Test | Unit Test | 1. Set annotation to `Prefer` <br> 2. Increase endpoints number to above `starting threshold` but below `starting threshold + padding` | `Prefer` will not be conducted | If capable, monitor the churn of EndpointSlices during the rolling update |
+|  |  | 1. Set annotation to `Prefer` <br> 2. Increase endpoints number to above `starting threshold + padding` | Distribution will be changed from `Balanced` to `Prefer` | Same as above |
+|  |  | 1. Set annotation to `Prefer` <br> 2. Decrease endpoints number to below `starting threshold` but above `starting threshold - padding` | Distribution will remain `Prefer` | Same as above |
+|  |  | 1. Set annotation to `Prefer` <br> 2. Decrease endpoints number to below `starting threshold - padding` | Distribution will be changed from `Prefer` to `Balanced` | Same as above |
+|  |  | 1. Set annotation to one of the three <br> 2. Set endpoints to above `starting threshold + padding` <br> 3. Manually change the annotation to another approach | Distribution will be changed based on the new value | Step 2 is only needed for `Prefer` |
+|  |  | 1. Set approach to `Balanced` <br> 2. Set endpoints to above `starting threshold + padding` <br> 3. Manually change the annotation to `Require` or `Prefer` <br> 4. Change the approach back to `Balanced` | Distribution will be changed from `Balanced` to `Require` or `Prefer` and changed back to `Balanced` | Step 2 is only needed for `Prefer` |
+| Rolling Update Test | Integration Test |1. Use integration framework to mock pods <br> 2. Set annotation to `Prefer` <br> 3. Conduct an aggressive rolling update policy for endpoints update <br> 4. Set endpoints to test corner cases | In the desired state, EndpointSlices are distributed as expected | If capable, monitor the churn of EndpointSlices during the rolling update |
+| Traffic Reliability Test | E2E Test |1. Set annotation to `Require` and `Prefer` <br> 2. Set endpoints above `starting threshold + padding` <br> 3. Reuse the current traffic reachability test | Endpoints are reachable from different zones as expected | Step 2 is only needed for `Prefer` |
+
+### Observability
```

Comment:
It would be good for administrators to know **which services** are "blowing out" of their zone. So, while it breaks tradition, I can't think of any way to do this besides a per-service metric.

This could also be an event, but that seems un-ergonomic.

Reply:
How would this metric be determined? If endpoints are being contributed from one zone to another for a Service?

Reply:
That's what I was thinking, yes. Otherwise you have no way of knowing which services need to be "re-balanced".  (I'm open to other suggestions - I just want to write an alert that says "service XXX may be costing you money")

Reply:
you can label the metrics with the zone name, so you can build a graph that shows the metrics per zone. 
I did something similar to the cidr_sets https://github.com/kubernetes/kubernetes/pull/90288/ , so you can have metrics per each cidr_set configured

Reply:
I agree this would be a valuable metric to have, but I think per-Service metrics may end up becoming a scalability issue. If we made these per-zone we'd potentially end up with data evening out across all Services in that zone even if there were significant discrepancies within each Service. Any ideas @wojtek-t?

Reply:
Would it be sufficient to note the number of imbalanced Services and then decorate the Service object itself (or EPSlices?) with an annotation or Condition to say "is imbalanced" ?